src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,700 @@
+        to self + other.
+        """
+        assert self._is_scalar_type(scalar)
+
+        if not inplace:
+            C = BlkDiagMatrix.from_blk_diag(self.data)
+        else:
+            C = self
+
+        for i in range(self.nblocks):
+            C[i] += scalar
+
+        return C
+
+    def sub(self, other):
+        """
+        Define the element subtraction of BlkDiagMatrix matrix
+
+        :param other: The rhs BlkDiagMatrix matrix
+
+        :return: A BlkDiagMatrix matrix with elementwise subraction equal to
+         self - other.
+        """
+        if self._is_scalar_type(other):
+            return self.scalar_sub(other)
+
+        self.check_compatible(other)
+
+        C = BlkDiagMatrix(self.nblocks, dtype=self.dtype)
+
+        for i in range(self.nblocks):
+            C[i] = self[i] - other[i]
+
+        return C
+
+    def __sub__(self, other):
+        """ Operator overloading for subtraction """
+        return self.sub(other)
+
+    def __isub__(self, other):
+        """ Operator overloading for in place subtraction """
+        if self._is_scalar_type(other):
+            return self.scalar_sub(other, inplace=True)
+        self.check_compatible(other)
+
+        for i in range(self.nblocks):
+            self[i] -= other[i]
+
+        return self
+
+    def __rsub__(self, other):
+        """ Convenient function for elementwise scalar subtraction """
+        # Note, the case of BlkDiagMatrix_L - BlkDiagMatrix_R would be
+        #   evaluated as L.sub(R), so this is only for other
+        #   Object - BlkDiagMatrix situations, namely scalars.
+        return -(self - other)
+
+    def scalar_sub(self, scalar, inplace=False):
+        """
+        Define the elementwise subtraction from BlkDiagMatrix matrix
+
+        :param scalar: constant subtractend value
+        :param inplace: bool, when false (default) return new instance.
+
+        :return:  BlkDiagMatrix matrix with elementwise sum equal to
+         self + other.
+        """
+        assert self._is_scalar_type(scalar)
+
+        if not inplace:
+            C = BlkDiagMatrix.from_blk_diag(self.data)
+        else:
+            C = self
+
+        for i in range(self.nblocks):
+            C[i] -= scalar
+
+        return C
+
+    def matmul(self, other):
+        """
+        Compute the Matrix multiplication of two BlkDiagMatrix matrices
+
+        :param other: The rhs BlkDiagMatrix matrix
+
+        :return: A BlkDiagMatrix of self @ other.
+        """
+        if not isinstance(other, BlkDiagMatrix):
+            raise RuntimeError(
+                "Attempt BlkDiagMatrix Matrix multiplication "
+                "(matmul,@) of non BlkDiagMatrix {}, try (*,mul)".format(
+                    repr(other)))
+
+        self.check_compatible(other)
+
+        C = BlkDiagMatrix(self.nblocks, dtype=self.dtype)
+
+        for i in range(self.nblocks):
+            C[i] = self[i] @ other[i]
+        return C
+
+    def __matmul__(self, other):
```

Comment:
Nope. No need. Matrix multiplication would require the left side is a BlkDiagMatrix, in which case python would implement `__matmul__` from the left.

The rhs corner cases are for this class only exist because we admit (scalar) types that are not BlkDiagMatrix objects.  I'm going to resolve this one, but obviously I can add some note if you want, just lmk next pass.